File: solution/1600-1699/1625.Lexicographically Smallest String After Applying Operations/Solution.go
func findLexSmallestString(s string, a int, b int) string {
	q := []string{s}
	vis := map[string]bool{s: true}
	ans := s
	for len(q) > 0 {
		s = q[0]
		q = q[1:]
		if s < ans {
			ans = s
		}
		for _, nxt := range []string{op1(s, a), op2(s, b)} {
			if !vis[nxt] {
				vis[nxt] = true
				q = append(q, nxt)
			}
		}
	}
	return ans
}

func op1(s string, a int) string {
	res := []byte(s)
	for i := 1; i < len(s); i += 2 {
		res[i] = byte((int(res[i]-'0')+a)%10 + '0')
	}
	return string(res)
}

func op2(s string, b int) string {
	return s[len(s)-b:] + s[:len(s)-b]
}

File: solution/1600-1699/1625.Lexicographically Smallest String After Applying Operations/Solution.ts
function findLexSmallestString(s: string, a: number, b: number): string {
    const q: string[] = [s];
    const vis = new Set<string>([s]);
    let ans = s;
    let i = 0;
    while (i < q.length) {
        s = q[i++];
        if (ans > s) {
            ans = s;
        }
        const t1 = s
            .split('')
            .map((c, j) => (j & 1 ? String((Number(c) + a) % 10) : c))
            .join('');
        const t2 = s.slice(-b) + s.slice(0, -b);
        for (const t of [t1, t2]) {
            if (!vis.has(t)) {
                vis.add(t);
                q.push(t);
            }
        }
    }
    return ans;
}
